Reaktoro/Core/Phases.hpp

```diff
@@ -1,387 +1,421 @@
+    }
+};
+
+/// The class used to configure a liquid solution phase.
+class LiquidPhase : public GenericPhase
+{
+public:
+    /// Construct a default LiquidPhase object.
+    LiquidPhase() : GenericPhase() { initialize(); }
+
+    /// Construct a LiquidPhase object with given species names.
+    explicit LiquidPhase(const StringList& species) : GenericPhase(species) { initialize(); }
+
+    /// Construct a LiquidPhase object with given element symbols.
+    explicit LiquidPhase(const Speciate& elements) : GenericPhase(elements) { initialize(); }
+
+    /// Construct an LiquidPhase object with given element symbols excluding the species with provided tags
+    explicit LiquidPhase(const Speciate& elements, const Exclude& withtags) { initialize(); };
+
+    /// Construct an LiquidPhase object excluding the species with provided tags
+    explicit LiquidPhase(const Exclude& withtags) { initialize(); };
+
+    /// Initialize the default attributes of this LiquidPhase object.
+    auto initialize() -> void
+    {
+        setName("LiquidPhase");
+        setStateOfMatter(StateOfMatter::Liquid);
+        setAggregateState(AggregateState::Liquid);
+        setActivityModel(ActivityModelIdealSolution());
+        setIdealActivityModel(ActivityModelIdealSolution());
+    }
+};
+
+/// The class used to configure a solid solution phase.
+class SolidPhase : public GenericPhase
+{
+public:
+    /// Construct a SolidPhase object with given species names.
+    explicit SolidPhase(const StringList& species) : GenericPhase(species) { initialize(); }
+
+    /// Initialize the default attributes of this SolidPhase object.
+    auto initialize() -> void
+    {
+        String phasename;
+        auto i = 0;
+        for(auto&& name : species())
+            phasename += (i++ == 0) ? name : "-" + name;
+
+        setName("SolidPhase");
+        setStateOfMatter(StateOfMatter::Solid);
+        setAggregateState(AggregateState::Solid);
+        setActivityModel(ActivityModelIdealSolution());
+        setIdealActivityModel(ActivityModelIdealSolution());
+    }
+};
+
+/// The class used to configure a pure mineral phase.
+class MineralPhase : public GenericPhase
+{
+public:
+    /// Construct a default MineralPhase object.
+    explicit MineralPhase(String mineral) : GenericPhase(mineral) { initialize(); }
+
+    /// Initialize the default attributes of this MineralPhase object.
+    auto initialize() -> void
+    {
+        setName(species().front());
+        setStateOfMatter(StateOfMatter::Solid);
+        setAggregateState(AggregateState::Solid);
+        setActivityModel(ActivityModelIdealSolution());
+        setIdealActivityModel(ActivityModelIdealSolution());
+    }
+};
+
+/// The class used to configure automatic selection of pure mineral phases.
+class MineralPhases : public GenericPhasesGenerator
+{
+public:
+    /// Construct a default MineralPhases object.
+    MineralPhases() : GenericPhasesGenerator() { initialize(); }
```

Comment:
We'll need an overloaded constructor for `MineralPhases` with an `Exclude` argument. This is because using `MineralPhases()` has the effect of collecting all minerals with chemical elements in the other phases. For example:

~~~c++
Phases phases(db);
phases.add( AqueousPhase(speciate("H O Na Cl C")) );
phases.add( GaseousPhase("H2O(g) CO2(g)") );
phases.add( MineralPhases() );  // This will result in all possible minerals with elements from the other phases added into `phases`.
~~~

We want to be able to allow:


~~~c++
Phases phases(db);
phases.add( AqueousPhase(speciate("H O Na Cl C")) );
phases.add( GaseousPhase("H2O(g) CO2(g)") );
phases.add( MineralPhases(exclude("carbonate")) );  // Say, assuming there are minerals with such tag `carbonate`
~~~

Reply:
Done